### PR TITLE
Adds enums to GDScript

### LINKED ISSUE #2966: Add enums to GDScript
### TL;DR
I would like to avoid writing
````js
const CORNER_W  = 0
const CORNER_NW = 1
const CORNER_NE = 2
const CORNER_E  = 3
const CORNER_SE = 4
const CORNER_SW = 5
````
And be able to write it like:
````js
enum {CORNER_W, CORNER_NW, CORNER_NE, CORNER_E, CORNER_SE, CORNER_SW}
````
----

### Some more ideas

Those ideas might simplify the usual workflow, or add additional features for strange workflows or setups.

Nice addition to avoid repetition:
````js
enum CORNER_{W, NW, NE, E, SE, SW}
````
Custom values idea 1:
````js
enum {CORNER_W = 1, CORNER_NW = 2, CORNER_NE = 3, CORNER_E = 4, CORNER_SE = 5, CORNER_SW = 6}
enum {CORNER_W = 4, CORNER_NW, CORNER_NE, CORNER_E, CORNER_SE, CORNER_SW}
````
Custom values idea 2:
````js
enum {CORNER_W, CORNER_NW, CORNER_NE, CORNER_E, CORNER_SE, CORNER_SW} = [1,2,3,4,5,6]
enum {CORNER_W, CORNER_NW, CORNER_NE, CORNER_E, CORNER_SE, CORNER_SW} = range(1,7)
````

Maybe `enum` should be prefixed by `const` or `var` depending on use? (IMO, it would be better to assume `const`, as to not create ambiguities, and/or hard to understand code)
````js
const enum {…}
var enum {…}
````

### PR DESCRIPTION
**Pull request by bojidar-bg**
Fixes #2966

Quick sample usage:

``` gdscript
enum {TLIE_FLOOR, TLIE_BRICK, TLIE_SPIKE, TLIE_MAX}
# TILE_FLOOR == 1, TILE_MAX == 4
enum {A = TLIE_BRICK,B,C}
# A==2, B==3, C==4
enum {PI_FLOORED = floor(PI), PI_CEILED = ceil(PI), PI_ROUNDED = round(PI)}
# Allows only ints though..
enum {TEST_A, TEST_B, TEST_C, # Also should allow newlines..
  TEST_A1 = TEST_A * 10, TEST_B1 = TEST_B * 10, TEST_C1 = TEST_C * 10}
# Because, why not (unintended consequence of the parsing code, heh)?
```

**DO NOT MERGE** I have to implement @djrm's proposal first.

### CODE REVIEW

**Comment by reduz** [comment]
I see you got used pretty fast to GDScript source code :P
I'm fine with enums, but the enum TILE_{} implementation while it sounds
cool for typing less, I would prefer users are forced to write the full
name. This will make it easier to eventually search/grep and code will be
more readable.

**Comment by kubecz3k** [comment]
@bojidar-bg woah enums ! <3 Are you planning to support them in expressions like `export (ENUMS) var name` maybe as well? :)

**Comment by bojidar-bg** [comment]
@kubecz3k I think, not yet, but open issue after this is merged :smile:

**Comment by punto-** [comment]
They should start from 0 shouldn't they? that way you can use "TILE_MAX" in
for loops or compare to array sizes, etc.

**Comment by PokemonAshLovesMyTurkeyAndILikeYouTwo** [comment]
what is the point of enums when we can just define that stuff in dict? (stupid q, I  know)

**Comment by bojidar-bg** [comment]
@punto- Fixed that one as well :smile: 
@PokemonAshLovesMyTurkeyAndILikeYouTwo Well, some people are really backwards when they come from a C++ background :wink: In fact, there is no difference, I think, in the final generated code...

**Comment by neikeq** [comment]
In terms of productivity, you don't need to type the value of every element. Also enum elements are constants, although I am not sure how constants work in GDScript.

**Comment by bojidar-bg** [comment]
@neikeq Nope, every would just get the last element's value + 1. Also, they are constants just fine :smile: 
@kubecz3k Ok, I did naming stuff, so that we can later on put the whole enum into an export.

**Comment by neikeq** [comment]
Oh, I forgot to ping @PokemonAshLovesMyTurkeyAndILikeYouTwo to my answer. It was about the comparison with dict.

**Comment by akien-mga** [comment]
Assigning to @reduz, as I'm not merging consequent GDScript changes :p

File: modules/gdscript/gd_parser.cpp [hunk 1]
```diff
@@ -3157,6 +3157,94 @@ void GDParser::_parse_class(ClassNode *p_class) {
...
+				if (tokenizer->get_token()!=GDTokenizer::TK_CURLY_BRACKET_OPEN) {
+					_set_error("Expected '{' in enum declaration");
+					return;
+				}
+				tokenizer->advance();
+				
+				while(true) {
+					if(tokenizer->get_token()==GDTokenizer::TK_NEWLINE) {
+						
+						tokenizer->advance(); // Ignore newlines
```
Comment (akien-mga) [inline modules/gdscript/gd_parser.cpp]: Would be great to have that in function parsing :D

**Comment by est31** [comment]
It would be cool if it would require the presence of a non empty common prefix so that all enums are namespaced.  C++ didn't do this until the introduction of enum class and you started wondering which enum some value belonged to and whether two values belong to the same enum (if you encountered a code base that didn't strictly execute the common prefix rule). Add that to missing static typing of the enum, and the confusion is perfect :).

**Comment by tomuxmon** [comment]
Should not syntax resemble python? Those curlies really look strange, (unless they [define a dictionary](http://stackoverflow.com/questions/9197324/curly-braces-in-python)).
Had a peak at [python syntax](https://www.python.org/dev/peps/pep-0435/)  and i it seems more natural and concise. I also like how[ enums and union types ](https://fsharpforfunandprofit.com/posts/enum-types/)are expressed in f# but I guess that does not conform to GDScript at all :)

Also I would give anything to have named enums and dot in to specific value like :
Kingdom.Fungi

**Comment by djrm** [comment]
Yes i also think that enums should be namespaced, and i propose the following syntax

``` gdscript
enum TILE: BLOCK, BRICK, WALL
enum TILE: (
   BLOCK,
   BRICK,
   WALL
)
```

**Comment by bojidar-bg** [comment]
@djrm Agreed, would do next week when I'm back at last :smile:

**Comment by reduz** [comment]
I would rather we don't namespace enums. This is intended to be a small helper, we are not making a statically typed language. 

@bojidar-bg If we ignore namespacing, is this PR ready to merge?

**Comment by reduz** [comment]
seems fine as is, so merging

**Comment by reduz** [comment]
someone should update GDScript docs on this!

**Comment by bojidar-bg** [comment]
@reduz Done :tada: https://github.com/godotengine/godot-docs/pull/248

**Comment by MarianoGnu** [comment]
First value is 1 by default? All languages i know start at 0

**Comment by akien-mga** [comment]
> First value is 1 by default? All languages i know start at 0

First value is 0.

**Comment by bojidar-bg** [comment]
@MarianoGnu Sorry, forgot to change first comment after my changes... guess I'll leave it like this.

**Pull request merged**